File: runner/scripts/lib/remote_cleanup.js
"use strict";

// Shared fail-closed result contract for pages/targets created through a
// remote task connection. Cleanup evidence belongs to the adapter outcome so
// the runner can persist it before deciding whether another attempt is safe.

function withCleanupObservations(outcome, cleanup) {
  const observations = {
    ...(outcome.observations || {}),
    target_cleanup: cleanup,
    isolation_restored: cleanup.confirmed === true,
  };
  return { ...outcome, observations };
}

function finalizeCleanupOutcome(outcome, cleanup, label, options = {}) {
  const recorded = withCleanupObservations(outcome, cleanup);
  const requireConfirmed = options.requireConfirmed !== false;
  if (!requireConfirmed || cleanup.confirmed === true) return recorded;
  return {
    ok: false,
    status: "infra",
    error: {
      class: "script_error",
      message: `${label} target cleanup was not confirmed: ${JSON.stringify(cleanup)}`,
    },
    answer: outcome.answer,
    observations: { ...recorded.observations, primary_outcome: outcome },
    metrics: outcome.metrics || {},
  };
}

function applyCleanupContract(outcome, cleanup, label) {
  return finalizeCleanupOutcome(outcome, cleanup, label, {
    requireConfirmed: true,
  });
}

function noTargetCleanup(backend) {
  return {
    backend,
    required: false,
    confirmed: true,
    same_connection_as_task: true,
    creation_attempts: [],
    attempts: [],
  };
}

module.exports = {
  applyCleanupContract,
  finalizeCleanupOutcome,
  noTargetCleanup,
  withCleanupObservations,
};

File: runner/scripts/adapters/remote_cleanup.py
"""Shared result contract for remote adapter-owned target cleanup."""

from __future__ import annotations

from typing import Any


def apply_cleanup_contract(
    outcome: dict[str, Any],
    cleanup: dict[str, Any],
    *,
    label: str,
) -> dict[str, Any]:
    observations = {
        **(outcome.get("observations") or {}),
        "target_cleanup": cleanup,
        "isolation_restored": cleanup.get("confirmed") is True,
    }
    if cleanup.get("confirmed") is True:
        return {**outcome, "observations": observations}
    return {
        "ok": False,
        "status": "infra",
        "error": {
            "class": "script_error",
            "message": f"{label} target cleanup was not confirmed: {cleanup!r}",
        },
        "answer": outcome.get("answer"),
        "observations": {**observations, "primary_outcome": outcome},
        "metrics": outcome.get("metrics") or {},
    }
